src/rust/rust-proto-new/src/graplinc/grapl/api/pipeline_ingress/v1beta1.rs

```diff
@@ -137,3 +137,199 @@ impl SerDe for PublishRawLogsResponse {
+    where
+        T: PipelineIngressApi<E> + Send + Sync + 'static,
+        E: ToString + Send + Sync + 'static
+    {
+        async fn publish_raw_logs(
+            &self,
+            request: Request<PublishRawLogsRequestProto>
+        ) -> Result<Response<PublishRawLogsResponseProto>, Status> {
+            let inner_request: PublishRawLogsRequest = match request.into_inner().try_into() {
+                Ok(request) => request,
+                Err(e) => return Err(Status::unknown(e.to_string())),
+            };
+
+            let response = self.api_server
+                .publish_raw_logs(inner_request)
+                .map_err(|e| Status::unknown(e.to_string()))
+                .await?;
+
+            let response: PublishRawLogsResponseProto = match response.try_into() {
+                Ok(res) => res,
+                Err(e) => return Err(Status::unknown(e.to_string())),
+            };
+
+            Ok(Response::new(response))
+        }
+    }
+
+    //
+    // public API
+    //
+
+    /// Implement this trait to define the pipeline ingress API's business logic
+    #[tonic::async_trait]
+    pub trait PipelineIngressApi<E>
+    where
+        E: ToString + 'static
+    {
+        async fn publish_raw_logs(
+            &self, request: PublishRawLogsRequest
+        ) -> Result<PublishRawLogsResponse, E>;
+    }
+
+    #[non_exhaustive]
+    #[derive(Debug, Error)]
+    pub enum ConfigurationError {
+        #[error("encountered tonic error {0}")]
+        TonicError(#[from] tonic::transport::Error),
+    }
+
+    /// The pipeline-ingress server serves the pipeline-ingress API
+    pub struct PipelineIngressServer<T, E>
+    where
+        T: PipelineIngressApi<E> + Send + Sync + 'static,
+        E: ToString + Send + Sync + 'static
+    {
+        api_server: T,
+        addr: SocketAddr,
+        e_: PhantomData<E>,
+    }
+
+    impl <T, E> PipelineIngressServer<T, E>
+    where
+        T: PipelineIngressApi<E> + Send + Sync + 'static,
+        E: ToString + Send + Sync + 'static
+    {
+        /// Construct a new gRPC server which will serve the given API
+        /// implementation on the given socket address. Server is constructed in
+        /// a non-running state. Call the serve() method to run the server.
+        pub fn new(api_server: T, addr: SocketAddr) -> Self {
+            PipelineIngressServer {
+                api_server,
+                addr,
+                e_: PhantomData
+            }
+        }
+
+        /// Run the gRPC server and serve the API on this server's socket
+        /// address. Returns a ConfigurationError if the gRPC server cannot run.
+        pub async fn serve(self) -> Result<(), ConfigurationError> {
+            Ok(Server::builder()
+               .add_service(PipelineIngressServiceServerProto::new(
+                   PipelineIngressProto::new(self.api_server)
+               ))
+               .serve(self.addr)
+               .map_err(|e| ConfigurationError::from(e))
```

Comment:
I think you can replace this with ?

Reply:
Yes this closure is redundant, will fix.